buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/transform/RestTestTransformer.java

```diff
@@ -0,0 +1,124 @@
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.gradle.test.rest.transform;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * Perform the transformations against the set of RestTests from a given file.
+ */
+public class RestTestTransformer {
+    /**
+     * Transforms a REST test based on the requested {@link RestTestTransform}'s
+     *
+     * @param tests           The REST tests from the same file. Uses linked list so we can easily add to the beginning of the list.
+     * @param transformations The set of transformations to perform against the test
+     * @return the transformed tests
+     */
+    public List<ObjectNode> transformRestTests(LinkedList<ObjectNode> tests, List<RestTestTransform<?>> transformations) {
+        ObjectNode setupSection = null;
+        ObjectNode teardownSection = null;
+
+        // Collect any global setup transformations
+        List<RestTestTransformGlobalSetup> setupTransforms = transformations.stream()
+            .filter(transform -> transform instanceof RestTestTransformGlobalSetup)
+            .map(transform -> (RestTestTransformGlobalSetup) transform)
+            .collect(Collectors.toList());
+
+        // Collect any global setup transformations
+        List<RestTestTransformGlobalTeardown> teardownTransforms = transformations.stream()
+            .filter(transform -> transform instanceof RestTestTransformGlobalTeardown)
+            .map(transform -> (RestTestTransformGlobalTeardown) transform)
+            .collect(Collectors.toList());
+
+        // transform the tests and include the global setup and teardown as part of the transform
+        for (ObjectNode test : tests) {
+            Iterator<Map.Entry<String, JsonNode>> testsIterator = test.fields();
+            while (testsIterator.hasNext()) {
+                Map.Entry<String, JsonNode> testObject = testsIterator.next();
+                String testName = testObject.getKey();
+                if ("setup".equals(testName)) {
+                    setupSection = test;
+                }
+                if ("teardown".equals(testName)) {
+                    teardownSection = test;
+                }
+                Map<String, RestTestTransformByObjectKey> objectKeyFinders = transformations.stream()
```

Comment:
this could be done before the `tests` for loop?

Reply:
indeed it can, nice catch.